### PR TITLE
PCgrad Implementation IC

### PR DESCRIPTION
**Pull request by jmcguigan10**

### CODE REVIEW

File: model_training/ml_loss.py [hunk 1]
```diff
@@ -40,4 +40,35 @@ def max_error(F4f_pred, F4f_true):
...
+                denom = sum((gj * gj).sum() for gj in grads_j)
+                if denom.item() > 0:
+                    scale = dot / (denom + eps)
+                    grads_proj = [gi - scale * gj for gi, gj in zip(grads_proj, grads_j)]
+        projected_grads.append(grads_proj)
+
+    merged_grads = []
+    for k in range(len(params)):
+        merged_grads.append(sum(grads[k] for grads in projected_grads))
+    return merged_grads
```
Comment (jmcguigan10) [inline model_training/ml_loss.py:71]: PCgrad Function Implementation. I will add some tests for this probably.

File: model_training/ml_pytorch.py [hunk 2]
```diff
@@ -45,6 +45,7 @@
     parms["weightedrandomsampler.epoch_num_samples"] = 10 #parms["samples_per_database"]
     
     parms["do_learn_task_weights"] = True
+    parms["do_pcgrad"] = False
```
Comment (jmcguigan10) [inline model_training/ml_pytorch.py]: Allow us to toggle pcgrad on or off.

File: model_training/ml_trainmodel.py [hunk 3]
```diff
@@ -61,6 +61,8 @@ def train_asymptotic_model(parms,
                            dataset_asymptotic_test_list,
                            dataset_stable_train_list,
                            dataset_stable_test_list):
+    if "do_pcgrad" not in parms:
+        parms["do_pcgrad"] = False
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py]: Just for the sake of being robust.

File: model_training/ml_trainmodel.py [hunk 4]
```diff
@@ -71,6 +73,7 @@ def train_asymptotic_model(parms,
     print("#Using",parms["device"],"device")
     if parms["device"] == "cuda":
         print("# ",torch.cuda.get_device_name(0))
+    print("#PCGrad", "on" if parms["do_pcgrad"] else "off")
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py]: Print out whether PCgrad is on or off.

File: model_training/ml_trainmodel.py [hunk 5]
```diff
@@ -162,26 +165,47 @@ def contribute_loss(pred, true, traintest, key, loss_fn):
...
-                                                                                                            (growthrate_true_train/ntot_invsec)) #torch.log
+            task_losses = []
+            task_losses.append(torch.exp(-model.log_task_weights["stability"] ) * stability_loss_fn(stable_pred_train, stable_true_train))
+            task_losses.append(torch.exp(-model.log_task_weights["ndens"]     ) * comparison_loss_fn(ndens_pred_train, ndens_true_train))
+            task_losses.append(torch.exp(-model.log_task_weights["fluxmag"]   ) * comparison_loss_fn(fluxmag_pred_train, fluxmag_true_train))
+            task_losses.append(torch.exp(-model.log_task_weights["direction"] ) *  direction_loss_fn(Fhat_pred_train, Fhat_true_train))
+            task_losses.append(torch.exp(-model.log_task_weights["growthrate"]) * comparison_loss_fn((growthrate_pred_train/ntot_invsec), #torch.log
+                                                                                                     (growthrate_true_train/ntot_invsec))) #torch.log
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py:171]: For PCgrad to work we need each task loss separate.

File: model_training/ml_trainmodel.py (hunk 3 above)
Comment (srichers) [reply model_training/ml_trainmodel.py]: I like to force parameters to be specified - I would rather make it easy to break loudly so we don't run into insipid issues based on not knowing what is happening under the hood.

File: model_training/ml_trainmodel.py [hunk 6]
```diff
@@ -162,26 +165,47 @@ def contribute_loss(pred, true, traintest, key, loss_fn):
...
 
             # add loss weights to loss
+            weight_loss = 0.0
             if parms["do_learn_task_weights"]:
                 for name in model.log_task_weights.keys():
                     if (not parms["do_unphysical_check"]) and name=="unphysical":
                         continue
                     else:
-                        batch_loss = batch_loss + torch.sum(model.log_task_weights[name])
+                        weight_loss = weight_loss + torch.sum(model.log_task_weights[name])
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py:182]: Just semantics, doesn't matter.

File: model_training/ml_trainmodel.py [hunk 7]
```diff
@@ -162,26 +165,47 @@ def contribute_loss(pred, true, traintest, key, loss_fn):
...
+                    param.grad = grad
+
+                if weight_params:
+                    total_loss = sum(task_losses) + weight_loss
+                    weight_grads = torch.autograd.grad(total_loss, weight_params, retain_graph=False, allow_unused=True)
+                    for param, grad in zip(weight_params, weight_grads):
+                        param.grad = grad if grad is not None else torch.zeros_like(param)
+            else:
+                batch_loss = sum(task_losses) + weight_loss
+                batch_loss.backward()
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py:205]: I think this should be right. The weighted loss shama and PCgrad should work fine together, I just thought tackling it on a case by case basis would be best.

Comment (jmcguigan10) [reply model_training/ml_trainmodel.py:205]: schema*

**Review by jmcguigan10 — COMMENTED** [review]
PCgrad implementation. We should add a Jenkins test to confirm pcgrad() function correctness.

File: model_training/ml_pytorch.py (hunk 2 above)
Comment (srichers) [reply model_training/ml_pytorch.py]: Can you turn pcgrad on by default so it gets run in Jenkins?

File: model_training/ml_trainmodel.py (hunk 4 above)
Comment (srichers) [reply model_training/ml_trainmodel.py]: all of the parameters are printed into a file at the start of each training. I think we can remove this line.

File: model_training/ml_trainmodel.py (hunk 7 above)
Comment (srichers) [reply model_training/ml_trainmodel.py:205]: should the weight losses not also be part of pcgrad?

File: model_training/ml_loss.py [hunk 8]
```diff
@@ -40,4 +40,35 @@ def max_error(F4f_pred, F4f_true):
         return 0
     else:
         return torch.max(torch.abs(F4f_pred - F4f_true))
+
+def pcgrad(task_losses, params, eps=1e-12):
+    if not params:
+        return []
+    if len(task_losses) == 0:
+        return [torch.zeros_like(p) for p in params]
```
Comment (srichers) [inline model_training/ml_loss.py]: I would rather have assert statements than "graceful" failure. I want it to break as fast as possible if something doesn't make sense. We should never run into a situation where parms or task_losses is empty, and if we do, it is a signal that something is wrong. It should crash.

File: model_training/ml_trainmodel.py (hunk 7 above)
Comment (jmcguigan10) [reply model_training/ml_trainmodel.py:205]: It does to weight losses first.

File: model_training/ml_loss.py (hunk 8 above)
Comment (jmcguigan10) [reply model_training/ml_loss.py]: done

File: model_training/ml_loss.py (hunk 1 above)
Comment (srichers) [reply model_training/ml_loss.py:71]: why apply pcgrad only if they are pointing in opposite directions? Why not just apply it all the time?

File: model_training/ml_trainmodel.py (hunk 4 above)
Comment (jmcguigan10) [reply model_training/ml_trainmodel.py]: done

**Review by jmcguigan10 — COMMENTED** [review]
Made the edits you suggested.

File: model_training/ml_trainmodel.py [hunk 9]
```diff
@@ -61,6 +61,7 @@ def train_asymptotic_model(parms,
                            dataset_asymptotic_test_list,
                            dataset_stable_train_list,
                            dataset_stable_test_list):
+    assert "do_pcgrad" in parms
```
Comment (srichers) [inline model_training/ml_trainmodel.py]: This will get checked the first time the variable is used. I think it would make sense to just remove this line. If it stays, then it becomes a liability for refactoring, and simplicity is best when possible.

File: model_training/ml_trainmodel.py (hunk 7 above)
Comment (srichers) [reply model_training/ml_trainmodel.py:205]: to me it looks like pcgrad is applied only to task_losses and shared_params, excluding weight_loss and weight_params

**Review by srichers — CHANGES_REQUESTED** [review]
some small questions and minor changes requested

File: model_training/ml_loss.py (hunk 1 above)
Comment (jmcguigan10) [reply model_training/ml_loss.py:71]: If you had vectors with a positive dot product and applied the "pseudo" Gram-Schmidt projection, it would make their dot product negative.

File: model_training/ml_trainmodel.py (hunk 7 above)
Comment (jmcguigan10) [reply model_training/ml_trainmodel.py:205]: Here the weight parameters are calculated and then pcgrad is applied:
```python
  if parms["do_learn_task_weights"]:
      weight_params = list(model.log_task_weights.parameters())
      weight_param_ids = {id(p) for p in weight_params}
      shared_params = [p for p in model.parameters() if id(p) not in weight_param_ids]
  else:
      weight_params = []
      shared_params = list(model.parameters())

  projected_grads = pcgrad(task_losses, shared_params)
```
Other than that, I didn't actually change anything structurally.

**Review by jmcguigan10 — COMMENTED** [review]
Removed the assert statement you asked me to.

File: model_training/ml_loss.py (hunk 1 above)
Comment (srichers) [reply model_training/ml_loss.py:71]: it would make the dot product zero, wouldn't it?

**Comment by srichers** [comment]
Can you pull from main, fix conflicts, and test this branch, too?
